### PR TITLE
log: debug manifest cache

### PR DESCRIPTION
**Pull request by crenshaw-dev**
Some more logging. It was critical to debugging https://github.com/argoproj/argo-cd/pull/9601, and I think it might be useful in the future.

### CODE REVIEW

File: reposerver/cache/cache.go [hunk 1]
```diff
@@ -189,10 +221,16 @@ func (c *Cache) SetManifests(revision string, appSrc *appv1.ApplicationSource, c
...
 	}
 
+	if log.IsLevelEnabled(log.DebugLevel) {
+		logEntryWithManifestCacheKeyFields(revision, appSrc, clusterInfo, namespace, trackingMethod, appLabelKey, appName, setReason).Debugf("setting manifests cache")
+	}
 	return c.cache.SetItem(manifestCacheKey(revision, appSrc, namespace, trackingMethod, appLabelKey, appName, clusterInfo), res, c.repoCacheExpiration, res == nil)
 }
 
-func (c *Cache) DeleteManifests(revision string, appSrc *appv1.ApplicationSource, clusterInfo ClusterRuntimeInfo, namespace string, trackingMethod string, appLabelKey string, appName string) error {
+func (c *Cache) DeleteManifests(revision string, appSrc *appv1.ApplicationSource, clusterInfo ClusterRuntimeInfo, namespace, trackingMethod, appLabelKey, appName, deleteReason string) error {
```
Comment (leoluz) [inline reposerver/cache/cache.go:230]: This function is already having lots of parameters. I don't think it is a good idea to introduce an additional one just to be used in the log. It is better to log with the specific reason before calling this function.

File: reposerver/cache/cache.go [hunk 2]
```diff
@@ -177,7 +209,7 @@ func (c *Cache) GetManifests(revision string, appSrc *appv1.ApplicationSource, c
 	return nil
 }
 
-func (c *Cache) SetManifests(revision string, appSrc *appv1.ApplicationSource, clusterInfo ClusterRuntimeInfo, namespace string, trackingMethod string, appLabelKey string, appName string, res *CachedManifestResponse) error {
+func (c *Cache) SetManifests(revision string, appSrc *appv1.ApplicationSource, clusterInfo ClusterRuntimeInfo, namespace string, trackingMethod string, appLabelKey string, appName string, res *CachedManifestResponse, setReason string) error {
```
Comment (leoluz) [inline reposerver/cache/cache.go:212]: This function is already having lots of parameters. I don't think it is a good idea to introduce an additional one just to be used in the log. It is better to log with the specific reason before calling this function.

File: reposerver/cache/cache.go [hunk 3]
```diff
@@ -139,14 +150,35 @@ func (c *Cache) GetGitReferences(repo string, references *[]*plumbing.Reference)
...
+		"appSrc":      appSourceKeyJSON(appSrc),
+		"namespace":   namespace,
+		"trackingKey": trackingKey(appLabelKey, trackingMethod),
+		"appName":     appName,
+		"clusterInfo": clusterRuntimeInfoKeyUnhashed(clusterInfo),
+		"reason":      reason,
+	})
+}
+
+func (c *Cache) GetManifests(revision string, appSrc *appv1.ApplicationSource, clusterInfo ClusterRuntimeInfo, namespace string, trackingMethod string, appLabelKey string, appName string, res *CachedManifestResponse, getReason string) error {
```
Comment (leoluz) [inline reposerver/cache/cache.go:177]: This function is already having lots of parameters. I don't think it is a good idea to introduce an additional one just to be used in the log. It is better to log with the specific reason before calling this function.

**Review by leoluz — CHANGES_REQUESTED** [review]
Function signatures are already huge and we can avoid adding additional parameters to them if we log in the caller function.

File: reposerver/cache/cache.go [hunk 4]
```diff
@@ -139,11 +150,32 @@ func (c *Cache) GetGitReferences(repo string, references *[]*plumbing.Reference)
...
 		trackingKey = trackingMethod + ":" + trackingKey
 	}
-	return fmt.Sprintf("mfst|%s|%s|%s|%s|%d", trackingKey, appName, revision, namespace, appSourceKey(appSrc)+clusterRuntimeInfoKey(info))
+	return trackingKey
+}
+
+// LogDebugManifestCacheKeyFields logs all the information included in a manifest cache key. It's intended to be run
+// before every manifest cache operation to help debug cache misses.
+func LogDebugManifestCacheKeyFields(message string, reason string, revision string, appSrc *appv1.ApplicationSource, clusterInfo ClusterRuntimeInfo, namespace string, trackingMethod string, appLabelKey string, appName string) {
+	if log.IsLevelEnabled(log.DebugLevel) {
```
Comment (leoluz) [inline reposerver/cache/cache.go:168]: I don't think this check is required. 
Logrus already checks for this inside the `Debug` function:
https://github.com/sirupsen/logrus/blob/v1.8.1/logger.go#L195-L201

**Review by leoluz — CHANGES_REQUESTED** [review]

File: reposerver/cache/cache.go (hunk 4 above)
Comment (crenshaw-dev) [reply reposerver/cache/cache.go:168]: It's just a small performance enhancement. Calculating the fields involves some JSON marshaling that I'd rather avoid unless we know we're going to log.

**Review by leoluz — APPROVED** [review]
LGTM

**Pull request merged**